packages/next/src/client/components/static-generation-bailout.ts

```diff
@@ -8,7 +8,10 @@ export function staticGenerationBailout(reason: string): boolean | never {
     return true
   }
 
-  if (staticGenerationStore?.isStaticGeneration) {
+  if (
+    staticGenerationStore?.isStaticGeneration ||
+    staticGenerationStore?.dynamicShouldError
+  ) {
     staticGenerationStore.revalidate = 0
     const err = new DynamicServerError(reason)
```

Comment:
This will cause an error during development but it'll deopt during production builds. In the `dynamicShouldError` case it should throw a different error that is not caught during `next build`. Seems tests for that are missing.

Reply:
I haven't found any recoveries of DynamicServerError but there is a bit more going on, so there is probably just some catch-all hidden somewhere.
Anyway refactored to use plain old Error, bonus is a better error message.